### PR TITLE
Fix channel args cli

### LINKED ISSUE #10732: defaults is added when conda-forge is the only channel and `-c` is given.
## Current Behavior

I don't understand what is the purpose of the following lines:
https://github.com/conda/conda/blob/e37cf84a57f935c578cdcea6ea034c80d7677ccc/conda/base/context.py#L700-L706

This is adding the `defaults` channel when using calling `conda install package -c conda-forge` when `conda-forge` is the only channel in the `.condarc`.

_Originally posted by @ericpre in https://github.com/conda/conda/issues/10479#issuecomment-775519701_

### Steps to Reproduce

Make `conda-forge` the only channel and do the following
```
conda create -n numpy numpy-base -c conda-forge
```
This will install numpy from defaults when it should error out because `defaults` is not supposed to be used at all.

## Expected Behavior

`defaults` should not be added.

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
     active environment : base
    active env location : /home/isuru/miniconda3
            shell level : 1
       user config file : /home/isuru/.condarc
 populated config files : /home/isuru/.condarc
          conda version : 4.10.1
    conda-build version : 3.21.4
         python version : 3.7.3.final.0
       virtual packages : __linux=4.15.0=0
                          __glibc=2.27=0
                          __unix=0=0
                          __archspec=1=skylake
       base environment : /home/isuru/miniconda3  (writable)
      conda av data dir : /home/isuru/miniconda3/etc/conda
  conda av metadata url : https://repo.anaconda.com/pkgs/main
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/isuru/tmp
       envs directories : /home/isuru/miniconda3/envs
                          /home/isuru/.conda/envs
               platform : linux-64
             user-agent : conda/4.10.1 requests/2.24.0 CPython/3.7.3 Linux/4.15.0-144-generic ubuntu/18.04.5 glibc/2.27
                UID:GID : 1000:1000
             netrc file : /home/isuru/.netrc
           offline mode : False
```
</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
==> /home/isuru/.condarc <==
add_pip_as_python_dependency: False
auto_activate_base: True
channels:
  - conda-forge
show_channel_urls: True
repodata_fns:
  - current_repodata.json
  - repodata.json
unsatisfiable_hints: True
```
</p></details>

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```

```
</p></details>

cc @chenghlee

### PR DESCRIPTION
**Pull request by ericpre**
Fix #10732.

### CODE REVIEW

File: tests/base/test_context.py [hunk 1]
```diff
@@ -342,6 +342,65 @@ def test_threads(self):
...
+        # should be used along with the one specified
+        # In this test, the given channel in cli is different from condarc
+        # 'defaults' should not be added
+        reset_context((), argparse_args=AttrDict(channel=['other']))
+        string = dals("""
+        channels: ['conda-forge']
+        """)
+        rd = odict(testdata=YamlRawParameter.make_raw_parameters('testdata', yaml_round_trip_load(string)))
+        context._set_raw_data(rd)
+        assert context.channels == ('conda-forge', 'other')
```
Comment (ericpre) [inline tests/base/test_context.py:388]: These tests check the current behaviour and pass without this PR.

File: tests/base/test_context.py [hunk 2]
```diff
@@ -342,6 +342,65 @@ def test_threads(self):
...
+        # In this test, the given channel in cli is the same as in condarc
+        # 'defaults' should not be added
+        # See https://github.com/conda/conda/issues/10732
+        reset_context((), argparse_args=AttrDict(channel=['conda-forge']))
+        string = dals("""
+        channels: ['conda-forge']
+        """)
+        rd = odict(testdata=YamlRawParameter.make_raw_parameters('testdata', yaml_round_trip_load(string)))
+        context._set_raw_data(rd)
+        assert context.channels == ('conda-forge', )
```
Comment (ericpre) [inline tests/base/test_context.py:402]: This test checks for the fix.

**Comment by ericpre** [comment]
At first, it was not clear to me when the `'defaults'` channel need to be added but based on the conda documentation, I added to tests cover the different cases and it should be good now.

File: conda/base/context.py [hunk 3]
```diff
@@ -720,11 +720,23 @@ def channels(self):
         if self._argparse_args and 'channel' in self._argparse_args:
```
Comment (ericpre) [inline conda/base/context.py:720]: This `if` block is necessary to cover the case in `test_specify_channels_cli_adding_defaults_no_condarc`.

File: tests/base/test_context.py [hunk 4]
```diff
@@ -342,6 +342,65 @@ def test_threads(self):
...
+        """)
+        rd = odict(testdata=YamlRawParameter.make_raw_parameters('testdata', yaml_round_trip_load(string)))
+        context._set_raw_data(rd)
+        assert context.channels == ('defaults', 'conda-forge')
+
+    def test_specify_different_channels_cli_condarc(self):
+        # When the channel have been specified in condarc, these channels
+        # should be used along with the one specified
+        # In this test, the given channel in cli is different from condarc
+        # 'defaults' should not be added
```
Comment (jezdez) [inline tests/base/test_context.py:381]:
```suggestion
        """
        When the channel have been specified in condarc, these channels
        should be used along with the one specified
        In this test, the given channel in cli is different from condarc
        'defaults' should not be added
        """
```

File: tests/base/test_context.py [hunk 5]
```diff
@@ -342,6 +342,65 @@ def test_threads(self):
...
+
+    def test_specify_channels_cli_adding_defaults_no_condarc(self):
+        # When the channel haven't been specified in condarc, 'defaults'
+        # should be present when specifying channel in the cli
+        reset_context((), argparse_args=AttrDict(channel=['conda-forge']))
+        assert context.channels == ('conda-forge', 'defaults')
+
+    def test_specify_channels_cli_condarc(self):
+        # When the channel have been specified in condarc, these channels
+        # should be used along with the one specified
```
Comment (jezdez) [inline tests/base/test_context.py:368]:
```suggestion
        """
        When the channel have been specified in condarc, these channels
        should be used along with the one specified
        """
```

File: tests/base/test_context.py [hunk 6]
```diff
@@ -342,6 +342,65 @@ def test_threads(self):
...
+        string = dals("""
+        channels: ['defaults', 'conda-forge']
+        """)
+        rd = odict(testdata=YamlRawParameter.make_raw_parameters('testdata', yaml_round_trip_load(string)))
+        context._set_raw_data(rd)
+        assert context.channels == ('defaults', 'conda-forge')
+
+    def test_specify_channels_cli_adding_defaults_no_condarc(self):
+        # When the channel haven't been specified in condarc, 'defaults'
+        # should be present when specifying channel in the cli
```
Comment (jezdez) [inline tests/base/test_context.py:362]:
```suggestion
        """
        When the channel haven't been specified in condarc, 'defaults'
        should be present when specifying channel in the cli
        """
```

File: tests/base/test_context.py [hunk 7]
```diff
@@ -342,6 +342,65 @@ def test_threads(self):
             assert context.repodata_threads == 1
             assert context.execute_threads == 3
 
+    def test_channels_defaults(self):
+        # no channels provided in cli
+        reset_context(())
+        assert context.channels == ('defaults', )
+
+    def test_channels_defaults_condarc(self):
+        # no channels provided in cli
```
Comment (jezdez) [inline tests/base/test_context.py:351]:
```suggestion
        """
        Test when no channels provided in cli, but some in condarc
        """
```

File: tests/base/test_context.py [hunk 8]
```diff
@@ -342,6 +342,65 @@ def test_threads(self):
             assert context.repodata_threads == 1
             assert context.execute_threads == 3
 
+    def test_channels_defaults(self):
+        # no channels provided in cli
```
Comment (jezdez) [inline tests/base/test_context.py:346]:
```suggestion
        """
        Test when no channels provided in cli
        """
```

**Review by jezdez — CHANGES_REQUESTED** [review]
This looks good to me in general, here are just some suggestions around using docstrings in the test functions so the pytest traceback has the full context. Minor fixes around tuple syntax and docstring content as well.

**Comment by ericpre** [comment]
Thanks @jezdez for the review, I have applied your suggestions!

**Review by jezdez — APPROVED** [review]
Thanks @ericpre!

**Pull request merged**